etsin_finder/frontend/js/stores/view/qvain.js

```diff
@@ -74,6 +79,207 @@ export class Qvain {
+      deselectDir(theDir)
+    }
+    this._hierarchy = newHier
+  }
+
+  @action toggleSelectedDirectory = (dir, select) => {
+    const newHier = { ...this._hierarchy }
+    const flat = getDirectories(newHier)
+    const theDir = flat.find(d => d.directoryName === dir.directoryName)
+    theDir.selected = select
+    if (select) {
+      theDir.files.forEach(f => {
+        f.selected = false
+      })
+      const deselectOthers = (aDir) => {
+        aDir.selected = false
+        aDir.files.forEach(f => {
+          f.selected = false
+        })
+        aDir.directories.forEach(d => deselectOthers(d))
+      }
+      theDir.directories.forEach(d => deselectOthers(d))
+    }
+    this._hierarchy = newHier
+  }
+
+  @action getInitialDirectories = () => (
+    axios
+      .get(PROJECT_DIR_URL + this._selectedProject)
+      .then(res => {
+        this._hierarchy = Directory(res.data, undefined, false, false)
+        return this._hierarchy
+      })
+      .catch(e => {
+        console.log('Failed to acquire project root directory, error: ', e)
+      })
+  )
+
+  @action changeProject = (projectId) => {
+    this._selectedProject = projectId
+    return this.getInitialDirectories()
+  }
+
+  @action loadDirectory = (dirId, rootDir, callback) => {
+    const req = axios
+      .get(DIR_URL + dirId)
+      .then(res => {
+        const newDirs = [...rootDir.directories.map(d => (
+          d.id === dirId ? {
+            ...d,
+            directories: res.data.directories.map(newDir => Directory(
+              newDir,
+              d,
+              false,
+              false
+            )),
+            files: res.data.files.map(newFile => File(
+              newFile,
+              d,
+              false
+            )),
+          } : d
+        ))]
+        rootDir.directories = newDirs
+      })
+    if (callback) {
+      req.then(callback)
+    }
+  }
+
+  @action setDirFileSettings = (directory, useCategory, fileType) => {
+    const newHier = { ...this._hierarchy }
+    const theDir = getDirectories(newHier).find(d => d.directoryName === directory.directoryName)
+    theDir.fileCharacteristics = {
+      useCategory,
+      fileType
+    }
+    this._hierarchy = newHier
+  }
+
+  @action setInEdit = (selectedItem) => {
+    this._inEdit = selectedItem
+  }
+
+  @computed
+  get userProjects() {
+    return this._userProjects
+  }
+
+  @computed
+  get selectedProject() {
+    return this._selectedProject
+  }
+
+  @computed
+  get selectedFiles() {
+    return this._selectedFiles
+  }
+
+  @computed
+  get selectedDirectories() {
+    return this._selectedDirectories
+  }
+
+  @computed
+  get inEdit() {
+    return this._inEdit
+  }
+
+  @computed
+  get hierarchy() {
+    return this._hierarchy
+  }
+
+  @computed
+  get directories() {
+    return this._directories
+  }
+
+  @computed
+  get files() {
+    return this._files
+  }
+
+  @computed
+  get parentDirs() {
+    return this._parentDirs
+  }
 }
 
```

Comment:
Plain JS object constructors for transforming file API JSON to an internal schema. Helps with managing the mess that is modifying the directory hierarchy.